Add allow-list support to resolve and rpmtree

kellyma2 opened the pull request:
In addition to allowing a set of packages that are not allowed, we should allow the scope to be narrowed to a set of packages that is allowed.  This change adds the `--only-allow` option to both `resolve` and `rpmtree` commands to achieve this outcome.

kellyma2 committed:
Add allow-list support to resolve and rpmtree

In addition to allowing a set of packages that are not allowed, we
should allow the scope to be narrowed to a set of packages that is
allowed.  This change adds the `--only-allow` option to both `resolve`
and `rpmtree` commands to achieve this outcome.

wade-arista commented on cmd/resolve.go:
```diff
@@ -79,6 +80,7 @@ func NewResolveCmd() *cobra.Command {
 	resolveCmd.Flags().BoolVarP(&resolveopts.nobest, "nobest", "n", false, "allow picking versions which are not the newest")
 	resolveCmd.Flags().StringArrayVarP(&resolveopts.repofiles, "repofile", "r", []string{"repo.yaml"}, "repository information file. Can be specified multiple times. Will be used by default if no explicit inputs are provided.")
 	resolveCmd.Flags().StringArrayVar(&resolveopts.forceIgnoreRegex, "force-ignore-with-dependencies", []string{}, "Packages matching these regex patterns will not be installed. Allows force-removing unwanted dependencies. Be careful, this can lead to hidden missing dependencies.")
+	resolveCmd.Flags().StringArrayVar(&resolveopts.onlyAllowRegex, "only-allow", []string{}, "Packages matching these regex patterns may be installed. Allows scoping dependencies. Be careful, this can lead to hidden missing dependencies.")
```
"Allows scoping dependencies" is a bit vague

wade-arista commented on pkg/sat/sat.go:
```diff
@@ -105,9 +105,10 @@ func (r *Resolver) ticket() string {
+// expressions which denote packages which should be taken into account for solving the problem, but they
 // should then be ignored together with their requirements in the provided list of installed packages.
-func (r *Resolver) LoadInvolvedPackages(packages []*api.Package, ignoreRegex []string) error {
+// Additionally, we can take a list of packages that the maximum selection desired.
```
What is "that the maximum selection desired"? Maybe write that a different way and, assuming it refers to `allowRegex`, mention that it means `allowRegex`?

wade-arista commented on pkg/sat/sat.go:
```diff
@@ -116,16 +117,33 @@ func (r *Resolver) LoadInvolvedPackages(packages []*api.Package, ignoreRegex []s
-					logrus.Warnf("Package %v is forcefully ignored by regex '%v'.", pkg.String(), rex)
-					r.forceIgnoreWithDependencies[pkg.String()] = packages[i]
-					break
+					matchedAllow = true
```
If you can only ever set it to `true` then you should break here.

wade-arista commented on pkg/sat/sat.go:
```diff
@@ -116,16 +117,33 @@ func (r *Resolver) LoadInvolvedPackages(packages []*api.Package, ignoreRegex []s
 				}
 			}
+
+			if !matchedAllow {
+				packages[i].Format.Requires.Entries = nil
+				logrus.Warnf("Package %v does not match allow list", pkg.String())
+				r.forceIgnoreWithDependencies[pkg.String()] = packages[i]
+			} else {
+				for _, rex := range ignoreRegex {
+					if match, err := regexp.MatchString(rex, fullName); err != nil {
+						return fmt.Errorf("failed to match package with regex '%v': %v", rex, err)
+					} else if match {
+						packages[i].Format.Requires.Entries = nil
+						logrus.Warnf("Package %v is forcefully ignored by regex '%v'.", pkg.String(), rex)
+						r.forceIgnoreWithDependencies[pkg.String()] = packages[i]
+						break
+					}
+				}
+			}
+			
```
I think it's better if you do:

```go
if matchedAllow  {
    for _, rex := range ignoreRegex {
        // maybe set matchedAllow back to false based on the `ignoreRegex` entries.
        ...
    }
}

if !matchedAllow {
    packages[i].Format.Requires.Entries = nil
    logrus.Warnf("Package %v does not match allow list", pkg.String())
    r.forceIgnoreWithDependencies[pkg.String()] = packages[i]
}

deduplicated[pkg.String()] = packages[i]
```

It just flattens out the code a bit and makes the logic a bit more explict, avoiding the repetition of this `packages[i].Format.Requires.Entries = nil` code block

wade-arista commented on pkg/sat/sat_test.go:
```diff
@@ -1229,7 +1229,7 @@ func Test(t *testing.T) {
 				packages = append(packages, &repo.Packages[i])
 			}
-			err = resolver.LoadInvolvedPackages(packages, nil)
+			err = resolver.LoadInvolvedPackages(packages, nil, nil)
```
test case where it's not nil?

kellyma2 committed:
Refinements and fixes

This change tweaks a few bits in the earlier changes to make things a
little nicer:

* clean up the flag help

* fixes the comments for LoadInvolvedPackages

* abort the allowed matcher loop early if we get a match

* re-works the allow/ignore code structure to reduce duplication

wade-arista commented on pkg/sat/sat.go:
```diff
@@ -104,10 +104,10 @@ func (r *Resolver) ticket() string {
-// expressions which denote packages which should be taken into account for solving the problem, but they
-// should then be ignored together with their requirements in the provided list of installed packages.
-// Additionally, we can take a list of packages that the maximum selection desired.
+// LoadInvolvedPackages takes a list of all involved packages to install, a list of regular expressions
+// which denote packages which should be taken into account for solving the problem, but they should
+// then be ignored together with their requirements in the provided list of installed packages, and also
+// a list of regular expressions that may be used to limit the selection to matching packages.
```
and the award for "best run on sentence" goes to... :melting_face:

kellyma2 replied:
I mean... it was already pretty long :)
